### PR TITLE
Added UI for taxonomy terms

### PR DESCRIPTION
**Pull request by harumijang**

### CODE REVIEW

**Comment by harumijang** [comment]
Added interfaces for `taxonomy--person_type` so far.
Can test by hitting `taxonomy/term/3`, `taxonomy/term/4,` `taxonomy/term/6` if you have the acms demo content.

Will continue adding the rest of the taxonomy vocabulary tomorrow. 

image

image

File: starters/basic-starter/pages/[...slug].tsx [hunk 1]
```diff
@@ -21,16 +25,57 @@ const CONTENT_TYPES = [
...
 }
 
-export default function NodePage({ node, menus }: NodePageProps) {
+export default function NodePage({ node, menus, label, type }: NodePageProps) {
   if (!node) return null;
 
-  return (
+  return Array.isArray(node) ? (
+    <Layout title={label} menus={menus}>
+      {type === 'taxonomy_term--article_type' &&
```
Comment (harumijang) [inline starters/basic-starter/pages/[...slug].tsx:47]: I am checking the `type` so existing taxonomy terms that don't have any nodes will still render a page with the term shown on the page

File: starters/basic-starter/pages/[...slug].tsx [hunk 2]
```diff
@@ -21,16 +25,57 @@ const CONTENT_TYPES = [
...
   'node--person',
   'node--place',
+  'taxonomy_term--article_type',
+  'taxonomy_term--categories',
+  'taxonomy_term--event_type',
+  'taxonomy_term--person_type',
+  'taxonomy_term--place_type',
 ];
 
 interface NodePageProps extends LayoutProps {
```
Comment (lauriii) [inline starters/basic-starter/pages/[...slug].tsx:35]: Hmm, `NodePageProps` sounds specific to nodes. I think we need some refactoring to make this generic so that it makes sense for both, taxonomy terms and nodes.

Comment (lauriii) [reply starters/basic-starter/pages/[...slug].tsx:35]: Maybe we should convert some of the logic into components and then this would become "EntityPage" or something along the lines of that

File: starters/basic-starter/pages/[...slug].tsx [hunk 3]
```diff
@@ -112,9 +163,81 @@ export async function getStaticProps(
   }
 
   // Fetch the node/resource from Drupal.
-  const node = await drupal.getResourceFromContext<DrupalNode>(type, context, {
-    params: params.getQueryObject(),
-  });
+  let node;
+  if (type === 'taxonomy_term--person_type') {
+    node = await drupal.getResourceCollectionFromContext<DrupalNode[]>(
```
Comment (lauriii) [inline starters/basic-starter/pages/[...slug].tsx:168]: I think it's confusing that we use the `node` property both for a list of nodes and for a single node. I think we should change this something else and make it so that both taxonomy term and list of nodes get passed to the component.

File: starters/basic-starter/pages/[...slug].tsx [hunk 4]
```diff
@@ -77,6 +122,12 @@ export async function getStaticProps(
 
   const type = path.jsonapi.resourceName;
 
+  let label;
+  if (path.label) {
+    // Get the taxonomy term from the path.
+    label = path.label;
```
Comment (lauriii) [inline starters/basic-starter/pages/[...slug].tsx:128]: For consistency, we should probably get the title for taxonomy terms from the entity object like we do for nodes

**Review by lauriii — CHANGES_REQUESTED** [review]

File: starters/basic-starter/pages/[...slug].tsx (hunk 4 above)
Comment (harumijang) [reply starters/basic-starter/pages/[...slug].tsx:128]: What if there is an existing taxonomy term with no associated objects like 'Software' for example

image

File: starters/basic-starter/pages/[...slug].tsx [hunk 5]
```diff
@@ -21,24 +22,43 @@ const CONTENT_TYPES = [
...
+  taxonomy_term?: string;
+  type?: string;
 }
 
-export default function NodePage({ node, menus }: NodePageProps) {
-  if (!node) return null;
-
-  return (
+export default function EntityPage({
+  node,
```
Comment (lauriii) [inline starters/basic-starter/pages/[...slug].tsx:40]: Can we call this entity which type would be either `DrupalNode` or `DrupalTaxonomyTerm`?

File: starters/basic-starter/pages/[...slug].tsx [hunk 6]
```diff
@@ -21,24 +22,43 @@ const CONTENT_TYPES = [
...
+  node_list,
```
Comment (lauriii) [inline starters/basic-starter/pages/[...slug].tsx:41]: I'd call this something more generic. Maybe `additionalContent` could work?

File: starters/basic-starter/components/taxonomy/taxonomy-term.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+import { TaxonomyPerson } from './taxonomy--person_type';
+import { TaxonomyEvent } from './taxonomy--event_type';
+import { TaxonomyPlace } from './taxonomy--place_type';
+import * as React from 'react';
+
+export function TaxonomyTerm({ node_list, menus, taxonomy_term, type }) {
+  return (
+    <Layout title={taxonomy_term} menus={menus}>
+      {type === 'taxonomy_term--article_type' &&
+        node_list.every(
```
Comment (lauriii) [inline starters/basic-starter/components/taxonomy/taxonomy-term.tsx:12]: I think we could actually potentially remove this check and move this handling to the individual term components. This would probably simplify this enough so that we could remove this component completely and move it back to the slug.

File: starters/basic-starter/pages/[...slug].tsx (hunk 4 above)
Comment (lauriii) [reply starters/basic-starter/pages/[...slug].tsx:103]: The taxonomy term object should be still available, even if there aren't any associated entities.

File: starters/basic-starter/pages/[...slug].tsx [hunk 8]
```diff
@@ -111,32 +137,130 @@ export async function getStaticProps(
...
+      );
+      break;
+    }
+    case 'taxonomy_term--article_type': {
+      node_list = await drupal.getResourceCollectionFromContext<DrupalNode[]>(
+        'node--article',
+        context,
+        {
+          params: new DrupalJsonApiParams()
+            .addInclude([
```
Comment (lauriii) [inline starters/basic-starter/pages/[...slug].tsx:163]: We should probably sort all of the lists. Maybe sorting by the creation time would make sense (this would be same as what we do in Drupal)?

**Review by lauriii — CHANGES_REQUESTED** [review]

File: starters/basic-starter/pages/[...slug].tsx (hunk 4 above)
Comment (harumijang) [reply starters/basic-starter/pages/[...slug].tsx:103]: Ah yeah, I misunderstood your first comment.

File: starters/basic-starter/pages/[...slug].tsx (hunk 5 above)
Comment (harumijang) [reply starters/basic-starter/pages/[...slug].tsx:40]: Could you rephrase this? I don't understand

**Comment by harumijang** [comment]
Addressed feedback, just need clarification on this comment:

> Can we call this entity which type would be either `DrupalNode` or `DrupalTaxonomyTerm`?

**Comment by harumijang** [comment]
Confirmed the taxonomy term links used in the test with a new fresh install with demo content

File: starters/basic-starter/pages/[...slug].tsx [hunk 9]
```diff
@@ -13,30 +13,84 @@ import { NodePerson } from 'components/node--person';
...
+        <NodePlace node={entity as DrupalNode} />
+      )}
+      {entity.type === 'taxonomy_term--article_type' && (
+        <TaxonomyArticle
+          additionalContent={additionalContent as { nodes: DrupalNode[] }}
+          taxonomy_term={entity as DrupalTaxonomyTerm}
+        />
+      )}
+      {entity.type === 'taxonomy_term--categories' && (
+        <TaxonomyArticle
```
Comment (lauriii) [inline starters/basic-starter/pages/[...slug].tsx:71]: Should we explain with an inline comment, why are we rendering the categories with TaxonomyArticle? I'd be fine with removing support for categories since it makes the least sense without having support for including multiple content types

Comment (harumijang) [reply starters/basic-starter/pages/[...slug].tsx:71]: I am fine with removing it too since it's a demo

**Review by lauriii — APPROVED** [review]
